finite_difference/src/subdomain_mod.f90

```diff
@@ -0,0 +1,302 @@
+    ! then they should share the domain between them:
+    internal_width = NINT(REAL(xlen) / REAL(ntilex))
+    internal_height = NINT(REAL(ylen) / REAL(ntiley))
+
+    ! Integer arithmetic means that ntiley tiles of height idy might
+    ! actually span a height greater or less than N. If so, we try and
+    ! reduce the height of each row by just one until we've accounted
+    ! for the <jover> extra rows.
+    !nwidth = (ntiley-2)*(idy-2) + 2*(idy-1)
+    nwidth = ntiley * internal_height
+    if(nwidth > ylen)then
+       jover  = nwidth - ylen
+       junder = 0
+    else if(nwidth < ylen)then
+       jover  = 0
+       junder = ylen - nwidth
+    else
+       jover  = 0
+       junder = 0
+    end if
+    ! Ditto for x dimension
+    !nwidth = (ntilex-2)*(idx-2) + 2*(idx-1)
+    nwidth = ntilex * internal_width
+    if(nwidth > xlen)then
+       iover  = nwidth - xlen
+       iunder = 0
+    else if(nwidth < xlen)then
+       iover  = 0
+       iunder = xlen - nwidth
+    else
+       iover  = 0
+       iunder = 0
+    end if
+
+    if(print_tiles)then
+       WRITE(*,"('Tile width = ',I4,', tile height = ',I4)") &
+            internal_width, internal_height
+       WRITE(*,"('iover = ',I3,', iunder = ',I3)") iover, iunder
+       WRITE(*,"('jover = ',I3,', junder = ',I3)") jover, junder
+    end if
+
+    ith = 1
+    ! The starting point of the tiles in y
+    jval = 1
+
+    nvects_max = 0
+    nvects_min = 1000000
+    nvects_sum = 0
+    decomp%max_width  = 0
+    decomp%max_height = 0
+
+    if(print_tiles)WRITE(*,"(/'Sub-domains:')")
+
+    do jj = 1, ntiley, 1
+
+       ! Point to subdomain we are about to define
+       subdomain => decomp%subdomains(ith)
+
+       ! If necessary, correct the height of this tile row
+       if(jover > 0)then
+          height = internal_height - 1
+          jover = jover - 1
+       else if(junder > 0)then
+          height = internal_height + 1
+          junder = junder - 1
+       else
+          height = internal_height
+       end if
+
+       !  . . . . . . . .
+       !  o o o o o . . .
+       !  o h h h h
+       !  o h h h h
+       !  o h h i i
+       !  o h h i i
+       !  o h h i i
+       !  .
+       !  .
+
+       ! The starting point of the tiles in x
+       ival = 1
+
+       do ji = 1, ntilex, 1
+         
+          ! If necessary, correct the width of this tile column
+          if(iover > 0)then
+             width = internal_width - 1
+             iover = iover - 1
+          else if(iunder > 0)then
+             width = internal_width + 1
+             iunder = iunder - 1
+          else
+             width = internal_width
+          end if
+
+          subdomain => decomp%subdomains(ith)
+          subdomain%internal%xstart = hwidth+1
+          subdomain%internal%xstop = subdomain%internal%xstart+width-1
+          subdomain%internal%nx = subdomain%internal%xstop - &
+                                  subdomain%internal%xstart + 1
```

Comment:
Isn't ...%nx = width? That looks a lot simpler ;)

Reply:
Oh yes. Ahem. :-)